p2p/rating/peersRatingHandler.go

```diff
@@ -0,0 +1,238 @@
+func (prh *peersRatingHandler) DecreaseRating(pid core.PeerID) {
+	prh.mut.Lock()
+	defer prh.mut.Unlock()
+
+	prh.updateRatingIfNeeded(pid, decreaseFactor)
+}
+
+func (prh *peersRatingHandler) getOldRating(pid core.PeerID) (int32, bool) {
+	oldRating, found := prh.topRatedCache.Get(pid.Bytes())
+	if found {
+		oldRatingInt, _ := oldRating.(int32)
+		return oldRatingInt, found
+	}
+
+	oldRating, found = prh.badRatedCache.Get(pid.Bytes())
+	if found {
+		oldRatingInt, _ := oldRating.(int32)
+		return oldRatingInt, found
+	}
+
+	return defaultRating, found
+}
+
+func (prh *peersRatingHandler) updateRatingIfNeeded(pid core.PeerID, updateFactor int32) {
+	oldRating, found := prh.getOldRating(pid)
+	if !found {
+		// new pid, add it with default rating
+		prh.topRatedCache.Put(pid.Bytes(), defaultRating, int32Size)
+		return
+	}
+
+	decreasingUnderMin := oldRating == minRating && updateFactor == decreaseFactor
+	increasingOverMax := oldRating == maxRating && updateFactor == increaseFactor
+	shouldSkipUpdate := decreasingUnderMin || increasingOverMax
+	if shouldSkipUpdate {
+		return
+	}
+
+	newRating := oldRating + updateFactor
+	if newRating > maxRating {
+		newRating = maxRating
+	}
+
+	if newRating < minRating {
+		newRating = minRating
+	}
+
+	prh.updateRating(pid, oldRating, newRating)
+}
+
+func (prh *peersRatingHandler) updateRating(pid core.PeerID, oldRating, newRating int32) {
+	oldTier := computeRatingTier(oldRating)
+	newTier := computeRatingTier(newRating)
+	if newTier == oldTier {
+		if newTier == topRatedTier {
+			prh.topRatedCache.Put(pid.Bytes(), newRating, int32Size)
+		} else {
+			prh.badRatedCache.Put(pid.Bytes(), newRating, int32Size)
+		}
+
+		return
+	}
+
+	prh.movePeerToNewTier(newRating, pid)
+}
+
+func computeRatingTier(peerRating int32) string {
+	if peerRating >= defaultRating {
+		return topRatedTier
+	}
+
+	return badRatedTier
+}
+
+func (prh *peersRatingHandler) movePeerToNewTier(newRating int32, pid core.PeerID) {
+	newTier := computeRatingTier(newRating)
+	if newTier == topRatedTier {
+		prh.badRatedCache.Remove(pid.Bytes())
+		prh.topRatedCache.Put(pid.Bytes(), newRating, int32Size)
+	} else {
+		prh.topRatedCache.Remove(pid.Bytes())
+		prh.badRatedCache.Put(pid.Bytes(), newRating, int32Size)
+	}
+}
+
+// GetTopRatedPeersFromList returns a list of peers, searching them in the order of rating tiers
+func (prh *peersRatingHandler) GetTopRatedPeersFromList(peers []core.PeerID, minNumOfPeersExpected int) []core.PeerID {
+	prh.mut.Lock()
+	defer prh.mut.Unlock()
+
+	peersTopRated := make([]core.PeerID, 0)
+	defer prh.displayPeersRating(&peersTopRated, minNumOfPeersExpected)
+
+	isListEmpty := len(peers) == 0
+	if minNumOfPeersExpected < minNumOfPeers || isListEmpty {
+		return make([]core.PeerID, 0)
+	}
+
+	peersTopRated, peersBadRated := prh.splitPeersByTiers(peers)
+	if len(peersTopRated) < minNumOfPeersExpected {
+		peersTopRated = append(peersTopRated, peersBadRated...)
+	}
```

Comment:
should we append all `peersBadRated` to `peersTopRated` if no minimum expected peers?